File: Hard/Shortest Path in 1-2 Graph/shortest-path-in-12-graph.py
from collections import deque
class Solution:
    def shortestPath(self, V: int, src: int, dest: int, edges: list[list[int]]) -> int:
        # code here
        adj = [[] for _ in range(V+len(edges))]
        
        vis = [False] * (V+len(edges))
        newnode = V
        for u,v,w in edges:
            if w==1:
                adj[u].append(v)
                adj[v].append(u)
            else:
                adj[u].append(newnode)
                adj[newnode].append(u)
                
                adj[v].append(newnode)
                adj[newnode].append(v)
                
                newnode+=1
        
        q = deque()
        q.append(src)
        dist = 0
        
        while(len(q) != 0):
            s = len(q)
            
            while(s > 0):
                s-=1
                cur = q.popleft()
                
                if cur==dest:
                    return dist
                
                for v in adj[cur]:
                    if not vis[v]:
                        vis[v] = True
                        q.append(v)
                
            dist+=1
        
        return -1
          
        

File: Difficulty: Hard/Shortest Path in 1-2 Graph/shortest-path-in-12-graph.cpp
class Solution {
  public:
    int shortestPath(int V, int src, int dest, vector<vector<int>> &edges) {
        
        // Create adjacency list.
        // Maximum possible nodes = Original vertices + one dummy node per edge.
        vector<vector<int>> adj(V + edges.size());

        // Index of the next dummy node.
        int newnode = V;

        // Build the transformed graph.
        for (auto vec : edges) {
            int u = vec[0];
            int v = vec[1];
            int w = vec[2];

            // If edge weight is 1, keep it unchanged.
            if (w == 1) {
                adj[u].push_back(v);
                adj[v].push_back(u);
            }
            else {
                // Replace every weight-2 edge:
                //
                //      u ----(2)---- v
                //
                // with
                //
                //      u --(1)-- dummy --(1)-- v
                //
                // so that every edge in the new graph has weight 1.

                adj[u].push_back(newnode);
                adj[newnode].push_back(u);

                adj[v].push_back(newnode);
                adj[newnode].push_back(v);

                // Move to the next available dummy node.
                newnode++;
            }
        }

        // Standard BFS.
        queue<int> q;

        // Visited array for all original + dummy nodes.
        vector<bool> vis(newnode, false);

        // Start BFS from the source node.
        q.push(src);
        vis[src] = true;

        // Stores the current BFS level,
        // which is also the shortest distance from src.
        int dist = 0;

        while (!q.empty()) {

            // Number of nodes present in the current level.
            int s = q.size();

            // Process one complete BFS level.
            while (s--) {

                int cur = q.front();
                q.pop();

                // Destination reached.
                // Since BFS explores level by level,
                // the first time we reach the destination
                // is guaranteed to be the shortest distance.
                if (cur == dest)
                    return dist;

                // Visit all unvisited neighbours.
                for (int v : adj[cur]) {
                    if (!vis[v]) {
                        vis[v] = true;
                        q.push(v);
                    }
                }
            }

            // Move to the next BFS level.
            dist++;
        }

        // Destination is unreachable.
        return -1;
    }
};